expr/sort.go

```diff
@@ -114,13 +122,19 @@ func sortByBraces(metrics []*types.MetricData, part int, pattern string) {
 	}
 }
 
+// compile regex for sorting
+// it splits string by dots, but ignoring enclosed {} and []
+// it fixes bug when number of dots in pattern and metric can differ
+// causing out of index error in getParts() above
+var reSplitByDotsIgnoreBraces = regexp.MustCompile(`\s*(\{[^}]*\}|\[[^]]*\]|[^.]+)`)
```

Comment:
I can't tell how much this would impact performance in this particular case, but regexs are very slow in Go, especially complex ones like this.

From what I understand, this can be done without regex by using a simple processing function and it would be much faster. This would also eliminate the need for pre-compilation.

On top of that, in my subjective opinion, it would be much easier to understand if it was a function instead of a regex.

Reply:
@grzkv: you're probably right, but I'm doubt that I would be able to write function which will be faster than regex.

Reply:
Myeah. Will try to write that function, at least for readability.

Reply:
> doubt that I would be able to write function which will be faster than regex

If I get this correctly, this looks like an O(N) check. It will almost certainly be faster and clearer than regex. See also https://github.com/golang/go/issues/26623

The fact that we've agreed on making the regex local makes this less acute. Global regex was the main reason I was bothering you with this) Still, it would be nice to have.

Reply:
I wrote split function, please review.